node/src/mempool.rs

```diff
@@ -107,21 +111,35 @@ impl MempoolSrv {
         vm: &Arc<RwLock<VM>>,
         tx: &Transaction,
     ) -> Result<(), TxAcceptanceError> {
-        // VM Preverify call
-        if let Err(e) = vm.read().await.preverify(tx) {
-            Err(TxAcceptanceError::VerificationFailed(format!("{e:?}")))?;
-        }
-
         let tx_id = tx.id();
 
         // Perform basic checks on the transaction
         db.read().await.view(|view| {
-            // ensure transaction does not exist in the mempool
+            let count = view.txs_count();
+            if count >= self.conf.max_mempool_txn_count {
+                return Err(TxAcceptanceError::MaxTxnCountExceeded(count));
+            }
 
+            // ensure transaction does not exist in the mempool
             if view.get_tx_exists(tx_id)? {
                 return Err(TxAcceptanceError::AlreadyExistsInMempool);
             }
 
+            // ensure transaction does not exist in the blockchain
+            if view.get_ledger_tx_exists(&tx_id)? {
+                return Err(TxAcceptanceError::AlreadyExistsInLedger);
+            }
+
+            Ok(())
+        })?;
+
+        // VM Preverify call
+        if let Err(e) = vm.read().await.preverify(tx) {
+            Err(TxAcceptanceError::VerificationFailed(format!("{e:?}")))?;
+        }
+
+        // Try to add the transaction to the mempool
+        db.read().await.update(|db| {
```

Comment:
To preserve the atomicity of the transaction acceptance, you should reuse the same `db.read().await` acquired before.

This way we should prevent race condition between mempool and block acceptance

Reply:
On a second tough, if we want to have an exclusive lock here we should use `db.write()` instead of `db.read()`

Theoretically for the VM too IMO

Reply:
> To preserve the atomicity of the transaction acceptance, you should reuse the same db.read().await acquired before.

> This way we should prevent race condition between mempool and block acceptance

> On a second tough, if we want to have an exclusive lock here we should use db.write() instead of db.read()

> Theoretically for the VM too IMO

Acquiring `db.write` here will block any db operation in all other threads. We try to be lock-free as much as possible without having corrupted data. 

 As we don't expect high lock contention (per key) between mempool and chain (light)threads, there is an `OptimisticTransactionDB` used under the hood.

There might be some (rare) races in which the mempool is receiving the same transaction that is  being accepted in the `chain` but this can be addressed when we implement `mempool::on_idle` to clean up expired/accepted transactions. 

NB: The issue here is about `commiting` the `delete_tx` operation (in other words) calling it in write db txn. Any considerations about the `DB transaction model` together with the `DB lock policy` we have is pretty major thing so if you're not convinced let's have another issue to tackle this.

Reply:
> this can be addressed when we implement mempool::on_idle to clean up expired/accepted transactions.

Do we already have an issue?